src/analyses/octApron.ml

```diff
@@ -0,0 +1,182 @@
+        let formargs = Goblintutil.zip f.sformals args in
+        let arith_formals = List.filter (fun (x,_) -> isArithmeticType x.vtype) formargs in
+        List.iter (fun (v, e) -> D.substitute_var_with nd' (v.vname^"'") e) arith_formals;
+        let vars = List.map (fun (x,_) -> x.vname^"'") arith_formals in
+        D.remove_all_with nd' vars;
+        D.forget_all_with nd' [v.vname];
+        D.substitute_var_eq_with nd' "#ret" v.vname;
+        D.remove_all_with nd' ["#ret"];
+        A.unify Man.mgr nd nd'
+      | _ -> D.topE (A.env ctx.local)
+
+  let special ctx r f args =
+    if D.is_bot ctx.local then D.bot () else
+      begin
+        match LibraryFunctions.classify f.vname args with
+        | `Assert expression -> ctx.local
+        | `Unknown "printf" -> ctx.local
+        | `Unknown "__goblint_check" -> ctx.local
+        | `Unknown "__goblint_commit" -> ctx.local
+        | `Unknown "__goblint_assert" -> ctx.local
+        | `Malloc size -> 
+          (match r with
+            | Some lv ->
+              D.remove_all ctx.local [f.vname]
+            | _ -> ctx.local)
+        | `Calloc (n, size) -> 
+          (match r with
+            | Some lv ->
+              D.remove_all ctx.local [f.vname]
+            | _ -> ctx.local)
+        | `ThreadJoin (id,ret_var) -> 
+            let nd = ctx.local in
+            invalidate nd [ret_var];
+            nd
+        | `ThreadCreate _ -> ctx.local
+        | _ ->
+          begin
+            let st =
+              match LibraryFunctions.get_invalidate_action f.vname with
+              | Some fnc -> let () = invalidate ctx.local (fnc `Write  args) in ctx.local
+              | None -> D.topE (A.env ctx.local)
+            in
+              st      
+          end
+      end
+
+  let branch ctx e b =
+    if D.is_bot ctx.local then 
+      D.bot () 
+    else
+      let res = D.assert_inv ctx.local e (not b) in 
+      if D.is_bot res then raise Deadcode;
+      res
+
+  let return ctx e f =
+    if D.is_bot ctx.local then D.bot () else
+      
+      let nd = match e with
+        | Some e when isArithmeticType (typeOf e) -> 
+          let nd = D.add_vars ctx.local (["#ret"],[]) in 
+          let () = D.assign_var_with nd "#ret" e in
+          nd
+        | None -> D.topE (A.env ctx.local)
+        | _ -> D.add_vars ctx.local (["#ret"],[])
+      in
+      let vars = List.filter (fun x -> isArithmeticType x.vtype) (f.slocals @ f.sformals) in
+      let vars = List.map (fun x -> x.vname) vars in
+      D.remove_all_with nd vars;
+      nd
+
+  let body ctx f =
+    if D.is_bot ctx.local then D.bot () else
+      let vars = D.typesort f.slocals in
+      D.add_vars ctx.local vars
+
+  let assign ctx (lv:lval) e =
+    if D.is_bot ctx.local then D.bot () else
+      match lv with
+      | Var v, NoOffset when isArithmeticType v.vtype && (not v.vglob) -> 
+        D.assign_var_handling_underflow_overflow ctx.local v e
+      | _ -> D.topE (A.env ctx.local)
```

Comment:
Maybe one wants to handle more things here? As it is, any assignment to a variable that is not a local variable without offset results in the entire octagon going to `top`?

Also, how do you handle escaping local variables?

Reply:
One could, e.g., limit oneself to only locals of arithmetic type that do not have their addresses taken. Then one can safely ignore all other assigns.